### ISSUE TITLE
Not narrowing on TypedDict#get

### ISSUE THREAD

**Issue by A5rocks** [microsoft/pyright#2009]
**Describe the bug**

`typeddict.get('tag')` does not narrow when `'tag'` is a `Literal`. Note that mypy also does this (https://github.com/python/mypy/issues/10659) so I am unsure if this is intentional or a typeshed problem -- I don't see any `TypedDict` definition in typeshed though.

**To Reproduce**

Make two `TypedDict`s with a shared key (`tag`) and assign both two different literals. Now, use an if statement w/ `.get('tag')` to narrow. In that if statement, use `reveal_type`, reveals `Dict1 | Dict2`

**Expected behavior**

`reveal_type` should reveal only a single one.

**Screenshots or Code**

```python
from typing import TypedDict, Literal, Union, cast

class Blah(TypedDict):
    tag: Literal['a']

class Blah2(TypedDict):
    tag: Literal['b']

n = cast(Union[Blah, Blah2], 'example')

if n['tag'] == 'a':
    reveal_type(n)  # N: revealed type is Blah

if n.get('tag') == 'a':
    reveal_type(n)  # N: revealed type is Blah | Blah2
    # (should be narrowed as Blah)
```

**VS Code extension or command-line**
VSCode "Pylance" extension, version `2021.6.2`.

**Comment by erictraut** [comment]
You're correct, pyright doesn't currently handle type narrowing for this pattern. Here is a [list of the patterns](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards) that are supported.

Each new pattern that we support for conditional type narrowing adds complexity and potentially slows down analysis, so we're selective in which patterns we support. In particular, we need some evidence that it's a pattern that appears frequently in Python code. There are two signals we use to inform this decision: 1) we receive input from a significant number of users that they want support for the pattern, or 2) other Python type checkers have already added support this pattern.

This doesn't meet the second criteria. It could eventually meet the first, but I'm doubtful if it would because I don't think this is a common pattern, and there is a simple, straightforward workaround — namely, use `n['tag'] == 'a'`.

For that reason, I'm going to mark this as "as designed" and close it. We might reconsider it in the future if there is sufficient interest.
